server/services/store/sqlstore/subscriptions.go

```diff
@@ -0,0 +1,275 @@
+// See LICENSE.txt for license information.
+
+package sqlstore
+
+import (
+	"database/sql"
+
+	sq "github.com/Masterminds/squirrel"
+	"github.com/mattermost/focalboard/server/model"
+	"github.com/mattermost/focalboard/server/services/store"
+
+	"github.com/mattermost/mattermost-server/v6/shared/mlog"
+)
+
+func subscriptionFields() []string {
+	return []string{
+		"block_type",
+		"block_id",
+		"workspace_id",
+		"subscriber_type",
+		"subscriber_id",
+		"notified_at",
+		"create_at",
+		"delete_at",
+	}
+}
+
+func valuesForSubscription(sub *model.Subscription) []interface{} {
+	return []interface{}{
+		sub.BlockType,
+		sub.BlockID,
+		sub.WorkspaceID,
+		sub.SubscriberType,
+		sub.SubscriberID,
+		sub.NotifiedAt,
+		sub.CreateAt,
+		sub.DeleteAt,
+	}
+}
+
+func (s *SQLStore) subscriptionsFromRows(rows *sql.Rows) ([]*model.Subscription, error) {
+	subscriptions := []*model.Subscription{}
+
+	for rows.Next() {
+		var sub model.Subscription
+		err := rows.Scan(
+			&sub.BlockType,
+			&sub.BlockID,
+			&sub.WorkspaceID,
+			&sub.SubscriberType,
+			&sub.SubscriberID,
+			&sub.NotifiedAt,
+			&sub.CreateAt,
+			&sub.DeleteAt,
+		)
+		if err != nil {
+			return nil, err
+		}
+		subscriptions = append(subscriptions, &sub)
+	}
+	return subscriptions, nil
+}
+
+// createSubscription creates a new subscription, or returns an existing subscription
+// for the block & subscriber.
+func (s *SQLStore) createSubscription(db sq.BaseRunner, c store.Container, sub *model.Subscription) (*model.Subscription, error) {
+	sub.WorkspaceID = c.WorkspaceID
+
+	if err := sub.IsValid(); err != nil {
+		return nil, err
+	}
+
+	now := model.GetMillis()
+
+	subAdd := *sub
+	subAdd.NotifiedAt = now // notified_at set so first notification doesn't pick up all history
+	subAdd.CreateAt = now
+	subAdd.DeleteAt = 0
+
+	query := s.getQueryBuilder(db).
+		Insert(s.tablePrefix + "subscriptions").
+		Columns(subscriptionFields()...).
+		Values(valuesForSubscription(&subAdd)...)
+
+	if s.dbType == mysqlDBType {
+		query = query.Suffix("ON DUPLICATE KEY UPDATE delete_at = 0, notified_at = ?", now)
+	} else {
+		query = query.Suffix("ON CONFLICT (block_id,workspace_id,subscriber_id) DO UPDATE SET delete_at = 0, notified_at = ?", now)
+	}
+
+	if _, err := query.Exec(); err != nil {
+		s.logger.Error("Cannot create subscription",
+			mlog.String("block_id", sub.BlockID),
+			mlog.String("workspace_id", sub.WorkspaceID),
+			mlog.String("subscriber_id", sub.SubscriberID),
+			mlog.Err(err),
+		)
+		return nil, err
+	}
+	return &subAdd, nil
+}
+
+// deleteSubscription soft deletes the subscription for a specific block and subscriber.
+func (s *SQLStore) deleteSubscription(db sq.BaseRunner, c store.Container, blockID string, subscriberID string) error {
+	now := model.GetMillis()
+
+	query := s.getQueryBuilder(db).
+		Update(s.tablePrefix+"subscriptions").
+		Set("delete_at", now).
+		Where(sq.Eq{"block_id": blockID}).
+		Where(sq.Eq{"workspace_id": c.WorkspaceID}).
```

Comment:
Same redundant filter, and same for the other instances

Reply:
I'm not 100% sure the filter is redundant from a security perspective. We should discuss at dev meeting, and if agreed, I can remove these in a separate PR.